library/ssl_tls.c

```diff
@@ -5957,36 +5957,53 @@ void mbedtls_ssl_conf_curves( mbedtls_ssl_config *conf,
 {
     conf->curve_list = curve_list;
 }
-#endif
+#endif /* MBEDTLS_ECP_C */
 
 #if defined(MBEDTLS_X509_CRT_PARSE_C)
 int mbedtls_ssl_set_hostname( mbedtls_ssl_context *ssl, const char *hostname )
 {
-    size_t hostname_len;
+    /* Initialize to suppress unnecessary compiler warning */
+    size_t hostname_len = 0; 
 
-    if( hostname == NULL )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    /* Check if new hostname is valid before making any change to current one */
 
-    hostname_len = strlen( hostname );
+    if( hostname != NULL )
+    {
+	hostname_len = strlen( hostname );
 
-    if( hostname_len + 1 == 0 )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+	if( hostname_len > MBEDTLS_SSL_MAX_HOST_NAME_LEN )
+	    return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
 
-    if( hostname_len > MBEDTLS_SSL_MAX_HOST_NAME_LEN )
-        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    /* Now it's clear that we will overwrite the old hostname, so free it safely */
 
-    ssl->hostname = mbedtls_calloc( 1, hostname_len + 1 );
+    if( ssl->hostname != NULL )
+    {
+        mbedtls_zeroize( ssl->hostname, strlen( ssl->hostname ) );
```

Comment:
Is it necessary to zeroize the hostname? After all, this is supposed to be a public value, so it just seems a bit inefficient to zeroize it.